### DISCUSSION TITLE
How to Use TIM1_CH1N using stm32duino core?

### DISCUSSION THREAD

**Discussion by China-Mr-Hou** [stm32duino/Arduino_Core_STM32#2287] [Q&A] [answered]
Hi everyone, I'm using G474RE as a core of a FOC Driver. It's MOSFET Driver requires both high input and low input (half bridge). Originally it uses Timer 1 - Channel 1-3 as high inputs whileas CH1N-CH3N as low inputs, using HAL_TIMEx_PWMN_Start(&htim1, TIM_CHANNEL_1). Is it possible to use this function under Arduino Framework just like in MDK? I couldn't find any TIM_HandleTypeDef objects :(

**Comment by fpistm** [comment]
First did you read the wiki:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/HardwareTimer-library

If you want the handle you can use:
https://github.com/stm32duino/Arduino_Core_STM32/blob/0ca1f4ca3d5936cf10855d2cde37e44735425edf/cores/arduino/HardwareTimer.h#L179

Under the hood a pin is associated to a specific TIM channel, ex for TIM1_CH1:
https://github.com/stm32duino/Arduino_Core_STM32/blob/0ca1f4ca3d5936cf10855d2cde37e44735425edf/variants/STM32G4xx/G473R(B-C-E)T_G474R(B-C-E)T_G483RET_G484RET/PeripheralPins.c#L144C95-L144C104
and there are other.

**Reply by China-Mr-Hou** [reply]
Sorry to trouble, but it just doesn't work as required :(
I wrote the following code as the HardwareTimer-library wiki:
Including library header file and declaring a HardwareTimer object.
```c
#include <HardwareTimer.h>
HardwareTimer *tim1 = new HardwareTimer(TIM1);
``` 
Then in setup() function
```c
tim1->setMode(STM_PIN_CHANNEL(pinmap_function(PA_8, PinMap_PWM)), TIMER_OUTPUT_COMPARE_PWM1, PA8);
tim1->setPrescaleFactor(2);
tim1->setOverflow(20000, HERTZ_FORMAT);
tim1->setCaptureCompare(STM_PIN_CHANNEL(pinmap_function(PA_8, PinMap_PWM)), 10, PERCENT_COMPARE_FORMAT);
tim1->resume();
``` 
Then PA8 outputs a 10% duty cycle, 6.67kHz square wave instead of 20kHz, I've tried setting prescale factor to 1.

However, I still have no idea how to make PB13, a pin related to TIM_CH1N, output a reverse signal of what PA8 outputs.

I tried the following (added before tim1->setPrescaleFactor() ) :
```c
tim1->setMode(STM_PIN_CHANNEL(pinmap_function(PB_13, PinMap_PWM)), TIMER_OUTPUT_COMPARE_PWM2, PB13);
``` 
Then I saw a reversed square wave (90% duty cycle) while probing PB13 pin, but nothing via PA8 pin. PA8 only output a stable 3.3v.

Nevertheless, I tried adding the lines below instead:
```c
HAL_TIMEx_PWMN_Start(tim1->getHandle(), 0x00000000U);
``` 
Adding this original HAL function just made everything worse, with PA8 stable 3.3v and PB13 stable 0v.

Lastly I simply copied some code calling HAL function from a MDK project that functions correctly, which didn't help since I believe Timer is not set via this method.

Thanks again for everyone who helped me, I'm still struggling my way...

p.s. I tried analogWrite(PC13, value) before and it worked well with a fading LED. (PC13 is also a pin that related to TIM1_CH1N btw.)

**Reply by fpistm** [reply, marked as answer]
Hi @China-Mr-Hou 
I've tested with a Nucleo G474RE and the frequency is correct:
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/20641798/867a4583-8587-482c-83b9-d478b6862c75)
Same using the Generic target.

Don't know which board you use with a STM32G474RE, you probably use the generic G474RE variant and I guess you redefine the clock config? This can explain your issue as depending of your clock config the precision can be high or not.
By default the generic or the Nucleo variants are configured to the max freq 170MHz.

About having TIM1_CH1 and TIM1_CH1N together, this is not supported by the current HT implementation. I will provide a PR to allow this.
Here a workaround to be able to have it:
```C++
#include <HardwareTimer.h>
HardwareTimer *tim1 = new HardwareTimer(TIM1);

void setup() {
  uint32_t channel = STM_PIN_CHANNEL(pinmap_function(PA_8, PinMap_PWM));
  uint32_t channeln = STM_PIN_CHANNEL(pinmap_function(PB_13, PinMap_PWM));
  tim1->setMode(channel, TIMER_OUTPUT_COMPARE_PWM1, PA8);
  tim1->setMode(channeln, TIMER_OUTPUT_COMPARE_PWM1, PB13);
  tim1->setOverflow(20000, HERTZ_FORMAT);
  tim1->setCaptureCompare(channel, 10, PERCENT_COMPARE_FORMAT);
  tim1->resume();
  HAL_TIM_PWM_Start(tim1->getHandle(), 0); // Workaround
}

void loop() {
  // put your main code here, to run repeatedly:
}
```


Note that `tim1->setPrescaleFactor(2);` is useless as the `tim1->setOverflow()` with `HERTZ_FORMAT` will compute to have the best value.

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/20641798/f6243768-cb41-4825-b0e3-427c53e16513)

**Reply by China-Mr-Hou** [reply]
Thanks! It worked!
Anyway, I'm still confused why my output frequency is one third of the number in setOverflow. I personally believe it was the clock config (HSE Frequency, etc.) that made. btw, I successfully had 20kHz by setting Overflow to 60000 with HERTZ_FORMAT config.
If possible, could you explain why HAL_TIM_PWM_Start is used alone after tim1->resume() ? Isn't HAL_TIMEx_PWMN_Start is also required under this workaround method? Or isn't tim1->resume() has a similar function of starting PWM?

**Reply by fpistm** [reply]
If you check the resume code you will see that if a N channel is configured it is start and not the standard one. So the HAL_TIMEx_PWMN_Start is called and the 
HAL_TIM_PWM_Start not.

**Reply by China-Mr-Hou** [reply]
I've got that! I sincerely appreciate it!

**Reply by fpistm** [reply]
PR will come soon to be able to have both start if configured. About the frequency, well check your system core clock config.

**Reply by fpistm** [reply]
What is your HSE value?
Code assume by default it is 24MHz.
https://github.com/stm32duino/Arduino_Core_STM32/blob/c12103bc9f622d964e5f60278f43c1870616641c/system/STM32G4xx/stm32g4xx_hal_conf_default.h#L175

So if different you have to redefine it.

**Reply by China-Mr-Hou** [reply]
Oh, that's why! My board has an 8MHz crystal as HSE, which is apparently a third of the default value. I'll redefine HSE value in my code. Wish stm32duino project better and better!

**Reply by fpistm** [reply]
Hi @China-Mr-Hou 
I've made the PR to properly handle the regular and complementary channel. See #2294

**Reply by JingtaiChen** [reply]
您好，胡先生能加个联系方式吗？用以讨论关于STM32在ArduinoIDE上运行的技术问题。 @China-Mr-Hou
